Use Prettier

willdurand opened the pull request:
Fix #5312

---

This PR adds Prettier and applies it to our code base (`src/` and `tests/` folders, only JS files).

Here are the important changes:

- the `indent` ESLint rule has been disabled because it conflicts with
Prettier and we do not need it anymore (this is recommended by Prettier
in [eslint-config-prettier](https://github.com/prettier/eslint-config-prettier), which I did not use because no other rule conflicts
with our ESLint config)
- Prettier removes parentheses for trivial math operations, but I
decided to revert these changes (with `// prettier-ignore`) for clarity (there
are 3 or 4 occurrences)
- I have disabled the `jsx-indent` rule once in the `LanguageTools`
because Prettier formats `Array.fill` with JSX in it in a way that
ESLint does not like, but that's the only file where it was needed so I
did not disable the rule globally

kumar303 commented:
By setting `--trailing-comma all`, I'm concerned about this in the [docs](https://prettier.io/docs/en/options.html#trailing-commas):

> This requires node 8 or a transform.

Are we on Node 8 in production already? I didn't see a transform in here.

kumar303 commented:
Can you describe what the workflow will be? I don't see any changes to the readme (but github says it can't show me a table of contents so I'm not sure). What about the [pre-commit hook](https://prettier.io/docs/en/precommit.html)? Also, how do we enforce that all code going to master will be using prettier? I would think we'd need to fail in CI if we detect a diff.

willdurand commented:
> Ideally, I would like to have a CI job to enforce the use of prettier.

Looks like `yarn prettier --list-different` as a new job would do the trick!

kumar303 commented:
> I don't want Prettier to be seen as an optional tool

I agree

>  I think it is not the end of the world if some code does not get formatted with Prettier by accident and gets merged into master

I don't think we should allow code to land on master without going through `prettier`. As part of this patch, I'd like to see a CI job that fails for code that was not run through `prettier`. 

My main concern is if you run `prettier` for a pull request and it changes *old* code, it will make code review cumbersome. Adopting `prettier` will only work if all developers use it all the time.

willdurand commented:
> My main concern is if you run prettier for a pull request and it changes old code, it will make code review cumbersome. Adopting prettier will only work if all developers use it all the time.

That makes sense, I have added a new check on Travis CI (and rebased).

kumar303 commented on package.json:
```diff
@@ -24,6 +24,8 @@
     "stylelint": "stylelint --syntax scss **/*.scss",
     "lint": "npm run eslint && npm run stylelint",
+    "prettier": "prettier --single-quote --trailing-comma all --arrow-parens always --write '{src,tests}/**/*.js'",
+    "prettier-ci": "yarn prettier --list-different",
```
I'm hesitant to see this use `yarn` because nothing else in `package.json` uses it. I know that Travis installs yarn so it's fine but I think it would be better to stick with `npm run` mostly because we already decided to move away from yarn. One key difference (I think) is that you'll need to pass in args with a double dash like `npm run prettier -- --list-different`.

willdurand replied:
works for me

kumar303 commented on .travis.yml:
```diff
@@ -48,6 +48,9 @@ jobs:
       env: COMMAND=nsp-check
+    - stage:
+      <<: *node-prod
+      env: COMMAND=prettier-ci
```
Ugh. This is a problem because it takes a long time (16 secs for me) and this stage is run *sequentially*, not [in parallel](https://docs.travis-ci.com/user/build-stages/). We should run it in parallel but I don't think there is an easy way to fix that. Unless you have an idea, don't worry about it in this patch. I filed this bug: https://github.com/mozilla/addons-frontend/issues/5366

willdurand commented on src/core/components/Paginate/index.js:
```diff
@@ -64,11 +64,13 @@ export class PaginateBase extends React.Component {
       // the end.
     } else if (end - currentPage < showExtra) {
-      return makePageNumbers({ start: end - showPages + 1, end });
+      // prettier-ignore
```
readability yes.

kumar303 replied:
I think the only way to use prettier is without exceptions. If there's no option to do what we want, we should succumb. Putting an ignore comment above a line of code also hurts readability.

willdurand replied:
I am fine with removing those comments for trivial maths.

willdurand replied:
It also means disabling one more ESLint rule..

willdurand commented on package.json:
```diff
@@ -263,6 +265,7 @@
     "po2json": "^0.4.5",
     "postcss-loader": "2.1.4",
     "potools": "^0.2.0",
+    "prettier": "^1.13.5",
```
ah right

willdurand commented on package.json:
```diff
@@ -24,6 +24,8 @@
     "stylelint": "stylelint --syntax scss **/*.scss",
     "lint": "npm run eslint && npm run stylelint",
+    "prettier": "prettier --single-quote --trailing-comma all --arrow-parens always --write '{src,tests}/**/*.js'",
+    "prettier-ci": "yarn prettier --list-different",
```
we had `test-ci` already so...

kumar303 replied:
[Broken window theory](https://en.wikipedia.org/wiki/Broken_windows_theory) 😢

willdurand replied:
Given that the command will likely fail with a message + error code, do we really need something else than `-ci`. It is not going to list anything anymore I believe.

kumar303 commented:
This is an example of what you'll see in a CI failure:

<details>
<summary>output</summary>

```
$ yarn prettier-ci
yarn run v1.7.0
$ yarn prettier --list-different
$ prettier --single-quote --trailing-comma all --arrow-parens always --write '{src,tests}/**/*.js' --list-different
src/locale/af/amo.jssrc/locale/af/amo.js
src/locale/ar/amo.jssrc/locale/ar/amo.js
src/locale/ast/amo.jssrc/locale/ast/amo.js
src/locale/bg/amo.jssrc/locale/bg/amo.js
src/locale/bn_BD/amo.jssrc/locale/bn_BD/amo.js
src/locale/bs/amo.jssrc/locale/bs/amo.js
src/locale/ca/amo.jssrc/locale/ca/amo.js
src/locale/cak/amo.jssrc/locale/cak/amo.js
src/locale/cs/amo.jssrc/locale/cs/amo.js
src/locale/da/amo.jssrc/locale/da/amo.js
src/locale/dbl/amo.jssrc/locale/dbl/amo.js
src/locale/dbr/amo.jssrc/locale/dbr/amo.js
src/locale/de/amo.jssrc/locale/de/amo.js
src/locale/dsb/amo.jssrc/locale/dsb/amo.js
src/locale/el/amo.jssrc/locale/el/amo.js
src/locale/en_GB/amo.jssrc/locale/en_GB/amo.js
src/locale/en_US/amo.jssrc/locale/en_US/amo.js
src/locale/es/amo.jssrc/locale/es/amo.js
src/locale/eu/amo.jssrc/locale/eu/amo.js
src/locale/fa/amo.jssrc/locale/fa/amo.js
src/locale/fi/amo.jssrc/locale/fi/amo.js
src/locale/fr/amo.jssrc/locale/fr/amo.js
src/locale/fy_NL/amo.jssrc/locale/fy_NL/amo.js
src/locale/ga_IE/amo.jssrc/locale/ga_IE/amo.js
src/locale/he/amo.jssrc/locale/he/amo.js
src/locale/hsb/amo.jssrc/locale/hsb/amo.js
src/locale/hu/amo.jssrc/locale/hu/amo.js
src/locale/id/amo.jssrc/locale/id/amo.js
src/locale/it/amo.jssrc/locale/it/amo.js
src/locale/ja/amo.jssrc/locale/ja/amo.js
src/locale/ka/amo.jssrc/locale/ka/amo.js
src/locale/kab/amo.jssrc/locale/kab/amo.js
src/locale/ko/amo.jssrc/locale/ko/amo.js
src/locale/mk/amo.jssrc/locale/mk/amo.js
src/locale/mn/amo.jssrc/locale/mn/amo.js
src/locale/ms/amo.jssrc/locale/ms/amo.js
src/locale/nb_NO/amo.jssrc/locale/nb_NO/amo.js
src/locale/nl/amo.jssrc/locale/nl/amo.js
src/locale/nn_NO/amo.jssrc/locale/nn_NO/amo.js
src/locale/pl/amo.jssrc/locale/pl/amo.js
src/locale/pt_BR/amo.jssrc/locale/pt_BR/amo.js
src/locale/pt_PT/amo.jssrc/locale/pt_PT/amo.js
src/locale/ro/amo.jssrc/locale/ro/amo.js
src/locale/ru/amo.jssrc/locale/ru/amo.js
src/locale/sk/amo.jssrc/locale/sk/amo.js
src/locale/sl/amo.jssrc/locale/sl/amo.js
src/locale/sq/amo.jssrc/locale/sq/amo.js
src/locale/sv_SE/amo.jssrc/locale/sv_SE/amo.js
src/locale/te/amo.jssrc/locale/te/amo.js
src/locale/th/amo.jssrc/locale/th/amo.js
src/locale/tr/amo.jssrc/locale/tr/amo.js
src/locale/uk/amo.jssrc/locale/uk/amo.js
src/locale/ur/amo.jssrc/locale/ur/amo.js
src/locale/vi/amo.jssrc/locale/vi/amo.js
src/locale/zh_CN/amo.jssrc/locale/zh_CN/amo.js
src/locale/zh_TW/amo.jssrc/locale/zh_TW/amo.js
error Command failed with exit code 1.
```

</details>

There is nothing about this error that would tell me I need to run prettier on my files. Hmm, any ideas for an easy fix? Typically we'd do this thing by adding a custom script in `bin/`. The `bin/start-func-test-server.js` has some code for firing off shell commands. Maybe there is a more lightweight way to do it.

willdurand commented:
> There is nothing about this error that would tell me I need to run prettier on my files. Hmm, any ideas for an easy fix? Typically we'd do this thing by adding a custom script in bin/. The bin/start-func-test-server.js has some code for firing off shell commands. Maybe there is a more lightweight way to do it.

Ah doh! I forgot to add `|| echo 'run prettier'` in the npm script. That should do the trick.

kumar303 commented:
> That should do the trick.

This may not preserve the exit status but it's worth a try

willdurand commented:
oh noes, we avoid to pin dependencies and we even have a test for that..

willdurand commented:
> that links to some sections in the docs about editor integration and git hooks

We don't do this for other tools, so I only linked to the website.. I emphasized the fact that the tool must be run prior to submit a PR though.

willdurand commented:
Note: this PR has low-priority compared to #5354 (which I'd love to land for next push)

SeanPrashad commented on README.md:
```diff
@@ -169,6 +171,12 @@ function getAllAddons({ categoryId }: GetAllAddonsParams = {}) {
 
+### Prettier
+
+We Use [Prettier][] to automatically format our JavaScript code and stop all the
```
Did you mean to say "We use [Prettier][] ..."?

kumar303 replied:
Ah, yes, that was a typo

kumar303 commented:
@SeanPrashad sorry, I just saw your comments

> Pretty-Quick might be the solution we're looking for

Great idea! I've filed a bug for this: https://github.com/mozilla/addons-frontend/issues/5448

> Furthermore, we can automate Prettier on an IDE level

Since everyone uses different editors, we haven't added built-in support for this. The Prettier docs have a lot of great resources for every kind of editor. I use the Vim integration and it's great.
